sdk/identity/identity/src/credentials/managedIdentityCredential/imdsMsi.ts

```diff
@@ -0,0 +1,140 @@
+    const expires = +requestBody.expires_on * 1000;
+    logger.info(`IMDS using expires_on: ${expires} (original value: ${requestBody.expires_on})`);
+    return expires;
+  } else {
+    // If these aren't possible, use expires_in and calculate a timestamp
+    const expires = Date.now() + requestBody.expires_in * 1000;
+    logger.info(`IMDS using expires_in: ${expires} (original value: ${requestBody.expires_in})`);
+    return expires;
+  }
+}
+
+function prepareRequestOptions(resource?: string, clientId?: string): RequestPrepareOptions {
+  const queryParameters: any = {
+    resource,
+    "api-version": imdsApiVersion
+  };
+
+  if (clientId) {
+    queryParameters.client_id = clientId;
+  }
+
+  return {
+    url: imdsEndpoint,
+    method: "GET",
+    queryParameters,
+    headers: {
+      Accept: "application/json",
+      Metadata: true
+    }
+  };
+}
+
+export const imdsMsi: MSI = {
+  async isAvailable(
+    identityClient: IdentityClient,
+    resource: string,
+    clientId?: string,
+    getTokenOptions?: GetTokenOptions
+  ): Promise<boolean> {
+    const { span, options } = createSpan(
+      "ManagedIdentityCredential-pingImdsEndpoint",
+      getTokenOptions
+    );
+
+    const request = prepareRequestOptions(resource, clientId);
+
+    // This will always be populated, but let's make TypeScript happy
+    if (request.headers) {
+      // Remove the Metadata header to invoke a request error from
+      // IMDS endpoint
+      delete request.headers.Metadata;
+    }
+
+    request.spanOptions = options.tracingOptions && options.tracingOptions.spanOptions;
+
+    try {
+      // Create a request with a timeout since we expect that
+      // not having a "Metadata" header should cause an error to be
+      // returned quickly from the endpoint, proving its availability.
+      const webResource = identityClient.createWebResource(request);
+      webResource.timeout = (options.requestOptions && options.requestOptions.timeout) || 500;
+
+      try {
+        logger.info(`Pinging IMDS endpoint`);
+        await identityClient.sendRequest(webResource);
+      } catch (err) {
+        if (
+          (err instanceof RestError && err.code === RestError.REQUEST_SEND_ERROR) ||
+          err.name === "AbortError" ||
+          err.code === "ECONNREFUSED" || // connection refused
+          err.code === "EHOSTDOWN" // host is down
+        ) {
+          // If the request failed, or NodeJS was unable to establish a connection,
+          // or the host was down, we'll assume the IMDS endpoint isn't available.
+          logger.info(`IMDS endpoint unavailable`);
+          span.setStatus({
+            code: CanonicalCode.UNAVAILABLE,
+            message: err.message
+          });
+
+          // IMDS MSI unavailable.
+          return false;
+        }
+      }
+
+      // If we received any response, the endpoint is available
+      logger.info(`IMDS endpoint is available`);
+
+      // IMDS MSI available!
```

Comment:
If getToken raises a 400 error, it will bubble up to this point: https://github.com/Azure/azure-sdk-for-js/pull/11807#discussion_r515426059

Where we throw CredentialUnavailable!

How does that sound?

Reply:
I'm asking more information about this through Teams. In any case, the changes related to any retry policy will happen through a separate issue and PR.